File: stdlib/ocaml/decimal_internal.ml
open Catala_runtime

(* Toplevel round_to_decimal *)
let round_to_decimal : decimal -> integer -> decimal =
 fun m n ->
  if n = Z.zero then Q.of_bigint (round m)
  else
    let pow_10 = Z.(pow (of_int 10) (to_int (Z.abs n))) |> Q.of_bigint in
    if Z.(Compare.(n > zero)) then
      let r = round (Q.mul m pow_10) |> Q.of_bigint in
      Q.div r pow_10
    else
      let r = round (Q.div m pow_10) |> Q.of_bigint in
      Q.mul r pow_10

let () =
  Catala_runtime.register_module "Decimal_internal"
    ["round_to_decimal", Obj.repr round_to_decimal]
    "*external*"

File: stdlib/python/decimal_internal.py
from catala_runtime import *

def round_to_decimal2(variable: Decimal, n_decimal: Integer) -> Decimal:
    if n_decimal == Integer(0):
        return variable.round()
    pow_10 = Decimal(pow(10, abs(int(n_decimal))))
    if n_decimal > Integer(0):
        return Decimal(variable * pow_10).round() / pow_10
    else:
        return Decimal(variable / pow_10).round() * pow_10

def round_to_decimal(variable: Decimal, n_decimal: Integer) -> Decimal:
    r = round_to_decimal2(variable, n_decimal)
    return r
